core/auth/src/main/java/software/amazon/awssdk/auth/credentials/InstanceProfileCredentialsProvider.java

```diff
@@ -285,21 +310,39 @@ private boolean isInsecureFallbackDisabled() {
     }
 
     private String[] getSecurityCredentials(String imdsHostname, String metadataToken) {
+        String profile = resolvedProfile.get();
+        if (profile != null) {
+            return new String[]{profile};
+        }
+
+        String urlBase = getSecurityCredentialsResource();
         ResourcesEndpointProvider securityCredentialsEndpoint =
             StaticResourcesEndpointProvider.builder()
-                                           .endpoint(URI.create(imdsHostname + SECURITY_CREDENTIALS_RESOURCE))
+                                           .endpoint(URI.create(imdsHostname + urlBase))
                                            .headers(getTokenHeaders(metadataToken))
-                .connectionTimeout(Duration.ofMillis(this.configProvider.serviceTimeout()))
+                                           .connectionTimeout(Duration.ofMillis(this.configProvider.serviceTimeout()))
                                            .build();
 
-        String securityCredentialsList =
-            invokeSafely(() -> HttpResourcesUtils.instance().readResource(securityCredentialsEndpoint));
-        String[] securityCredentials = securityCredentialsList.trim().split("\n");
+        try {
+            String securityCredentialsList =
+                invokeSafely(() -> HttpResourcesUtils.instance().readResource(securityCredentialsEndpoint));
+            String[] securityCredentials = securityCredentialsList.trim().split("\n");
+
+            if (securityCredentials.length == 0) {
+                throw SdkClientException.builder().message("Unable to load credentials path").build();
+            }
+
+            apiVersion.compareAndSet(ApiVersion.UNKNOWN, ApiVersion.EXTENDED);
+            resolvedProfile.set(securityCredentials[0]);
+            return securityCredentials;
 
-        if (securityCredentials.length == 0) {
-            throw SdkClientException.builder().message("Unable to load credentials path").build();
+        } catch (Ec2MetadataClientException e) {
+            if (e.statusCode() == 404 && apiVersion.compareAndSet(ApiVersion.UNKNOWN, ApiVersion.LEGACY)) {
```

Comment:
So it just occurred to me that there might be an issue with just the atomic swap here.

If there are two threads that attempt to get credentials at the same time, T1 and T2, and apiVersion == UNKNOWN, and the instance doesn't support EXTENDED, both T1 and T2 will initially get a 404. BUT since `apiVersion.compareAndSet(ApiVersion.UNKNOWN, ApiVersion.LEGACY)` would only succeed for _either_ T1 or T2, one of them will get `Failed to load credentials from IMDS.` exception. Does that sound right?

Reply:
@zoewangg WDYT?

Reply:
Something like this might work 

```java
ApiVersion currentVersion = apiVersion.get();

if (currentVersion == UNKNOWN) {
    try {
        // get credentials from extended endpoint
        apiVersion.compareAndSet(UNKNOWN, EXTENDED);
    } catch (Ec2MetadataClientException e) {
        if (apiVersion.compareAndSet(UNKNOWN, LEGACY) || apiVersion.get() == LEGACY) {
          // get credentials from LEGACY
        } else {
          throw e;
        }
    }
} else if (currentVersion == EXTENDED) {
    // get credentials from extended or throw
} else {
    // get credentials from legacy or throw
}
```

Reply:
I'm still a little skeptical of using Atomics and compareAndSet at all here - `refreshCredentials` is a private method called only in a synchronized block (by the cached supplier).  I understand that its being defensive - but I think it also makes the code read as if its intended to be thread safe, but I think we generally wouldn't want to allow multiple concurrent imds http calls and we'd probably want to ensure the logic we're using for getting/using the token would work with concurrent calls....